src/main/flight/autopilot.c

```diff
@@ -91,14 +196,172 @@ void altitudeControl(float targetAltitudeCm, float taskIntervalS, float vertical
+            float ewDistance; // cm, steps of 11.1cm, East of target is positive
+            GPS_distances(&gpsSol.llh, &currentTargetLocation, &nsDistance, &ewDistance);
+            float distanceCm = sqrtf(sq(nsDistance) + sq(ewDistance));
+    
+            posHold.NS.distance = nsDistance;
+            posHold.EW.distance = ewDistance;
+    
+            posHold.pt1Gain = pt1FilterGain(posHold.lpfCutoff, posHold.gpsDataIntervalS);
+            const float leak = 1.0f - 0.4f * posHold.gpsDataIntervalS; // gpsDataIntervalS is not more than 1.0s
+    
+            // ** Sanity check **
+            // larger threshold if faster at start
+            if (posHold.NS.isStarting || posHold.EW.isStarting) {
+                posHold.sanityCheckDistance = gpsSol.groundSpeed > 1000 ? gpsSol.groundSpeed : 1000.0f;
+                // 1s of flight at current speed or 10m, in cm
+            }
+            // primarily to detect flyaway from no Mag or badly oriented Mag
+            // but must accept some overshoot at the start, especially if entering at high speed
+            if (distanceCm > posHold.sanityCheckDistance) {
+                return false; // must stay within 10m or probably flying away
+                // value at this point is a 'best guess' to detect IMU failure in the event the user has no Mag
+                // if entering poshold from a stable hover, we would only exceed this if IMU was disoriented
+                // if entering poshold at speed, it may overshoot this value and falsely fail, if so need something more complex
+            }
+    
+            vectors_t *vectors[] = { &posHold.NS, &posHold.EW };
+            for (int i = 0; i < 2; i++) {
+                vectors_t *latLong = vectors[i];
+    
+                // separate PID controllers for latitude (NorthSouth or ns) and longitude (EastWest or ew)
+    
+                // ** P **
+                float pidP = latLong->distance * positionPidCoeffs.Kp;
+    
+                // ** I **
+                if (!latLong->isStarting){
+                    // only accumulate iTerm after completing the start phase
+                    // perhaps need a timeout on the start phase ?
+                    latLong->integral += latLong->distance * posHold.gpsDataIntervalS;
+                } else {
+                    // while moving sticks, slowly leak iTerm away, approx 2s time constant
+                    latLong->integral *= leak;
+                }
+                float pidI = latLong->integral * positionPidCoeffs.Ki;
+        
+                // ** D ** //
+                // get change in distance in NS and EW directions from gps.c using the `GPS_distances` function
+                // this gives cleaner velocity data than the module supplied GPS Speed and Heading information
+                float velocity = (latLong->distance - latLong->previousDistance) * posHold.gpsDataFreqHz; // cm/s, minimum step 11.1 cm/s
```

Comment:
step can be 11cm * 100Hz. And longitude delta is smaller near poles.
And it is implementation detail, GPS can provide better precision

Reply:
GPS serial comms are so slow that 20Hz is the practical limit of data points.
I noted the step size only because I saw these curious steps in the log.